src/idl_parser.cpp

```diff
@@ -3210,12 +3226,14 @@ CheckedError Parser::DoParse(const char *source, const char **include_paths,
         return Error("unable to locate include file: " + name);
       if (source_filename)
         files_included_per_file_[source_filename].insert(filepath);
-      if (included_files_.find(filepath) == included_files_.end()) {
+
+      std::string contents;
+      bool file_loaded = LoadFile(filepath.c_str(), true, &contents);
+      if (included_files_.find(HashFile(filepath.c_str(), contents.c_str())) ==
+          included_files_.end()) {
         // We found an include file that we have not parsed yet.
-        // Load it and parse it.
-        std::string contents;
-        if (!LoadFile(filepath.c_str(), true, &contents))
-          return Error("unable to load include file: " + name);
+        // Parse it.
+        if (!file_loaded) return Error("unable to load include file: " + name);
```

Comment:
The same, why `file_loaded` checked if and only its hash not found?
Why empty `contents`  is used for calculation?

Reply:
When we encounter an include directive, we need to answer two questions:
  1. Has the included already been parsed?
  2. If not, can we load it for parsing?
  
It's necessary to defer query 2 until we know the answer to query 1 is "no", becasue `Parser::Parse()` can be called schemas that do not exist on-disk ([example](https://github.com/google/flatbuffers/blob/1da6f4f18bdc7f47cd84f58a427835f2b790cb29/tests/test.cpp#L1305)). If one of those in-memory schemas is subsequently included in another schema, we won't be able to load it from disk but we can know that it was previously parsed based on its name alone. At least, that is the assumtion that was made even prior to  #6371.





I believe the main case where this happens is when you pass multiple files into one invocation of flatc.

Reply:
@mmmspatz thank you for the explanation.
Your solution looks correct.